optuna/study.py

```diff
@@ -510,13 +510,16 @@ def _run_trial(self, func, catch):
             self.logger.info(message)
             self._storage.set_trial_state(trial_id, structs.TrialState.PRUNED)
             return trial
-        except catch as e:
+        except Exception as e:
             message = 'Setting status of trial#{} as {} because of the following error: {}'\
                 .format(trial_number, structs.TrialState.FAIL, repr(e))
             self.logger.warning(message, exc_info=True)
             self._storage.set_trial_system_attr(trial_id, 'fail_reason', message)
             self._storage.set_trial_state(trial_id, structs.TrialState.FAIL)
-            return trial
+
+            if any(isinstance(e, cls) for cls in catch):
```

Comment:
According to [the reference](https://docs.python.org/3/library/functions.html#isinstance), `isinstance` accepts a tuple of exceptions as a second argument. I think we can simplify this line as follows:
```suggestion
            if isinstance(e, catch):
```

But, I have a concern about this implementation. It does not accept a list of exceptions although some novice users may not care about the difference between `tuple` and `list`.
What do you think of it?

Reply:
The current type hint of `catch` argument is `Union[Tuple[()], Tuple[Type[Exception]]]`, so, IMO, we shouldn't accept `list`.
I think that it may be better to add a type check to prevent users from passing a list unintendedly.
(or change the type hint to accept list type).

Reply:
I agree with @sile.

> add a type check to prevent users from passing a list unintendedly.

That's a good idea. It would be helpful for novice users of Python to fix their code.

Reply:
Added a type check in `optimize` (to keep the callstack shallow when raised).